lib/python/bind_data_array.h

```diff
@@ -64,6 +64,65 @@ void bind_common_mutable_view_operators(pybind11::class_<T, Ignored...> &view) {
       });
 }
 
+template <class D> auto cast_to_dict_key(const py::handle &obj) {
+  using key_type = typename D::key_type;
+  if constexpr (std::is_same_v<key_type, std::string>) {
+    return obj.cast<std::string>();
+  } else {
+    return key_type{obj.cast<std::string>()};
+  }
+}
+
+template <class D> auto cast_to_dict_value(const py::handle &obj) {
+  using val_type = typename D::mapped_type;
+  return obj.cast<val_type>();
+}
+
+template <class T, class... Ignored, class Set>
+void bind_dict_update(pybind11::class_<T, Ignored...> &view, Set &&set_item) {
+  view.def(
+      "update",
+      [set_item](T &self, const py::object &other, const py::kwargs &kwargs) {
+        if (!other.is_none()) {
+          if (py::hasattr(other, "keys")) {
+            // other is dict-like
+            for (const auto &key : other.attr("keys")()) {
+              set_item(self, cast_to_dict_key<T>(key),
+                       cast_to_dict_value<T>(other[key]));
+            }
+          } else {
+            // other is sequence of tuples
+            for (const auto &item : other) {
+              const auto &t = item.cast<py::tuple>();
+              set_item(self, cast_to_dict_key<T>(t[0]),
+                       cast_to_dict_value<T>(t[1]));
+            }
+          }
+        }
+
+        for (const auto &item : kwargs) {
+          set_item(self, cast_to_dict_key<T>(item.first),
+                   cast_to_dict_value<T>(item.second));
+        }
```

Comment:
Is this behaving consistently (with `dict`) when the same keys are given in the dict and kwargs?

Reply:
Yes. In fact the careful reader might find some overlap in the docstrings :)
```
update(...)
 D.update([E, ]**F) -> None.  Update D from dict/iterable E and F.
 If E is present and has a .keys() method, then does:  for k in E: D[k] = E[k]
 If E is present and lacks a .keys() method, then does:  for k, v in E: D[k] = v
 In either case, this is followed by: for k in F:  D[k] = F[k]
```